Assets/Mirror/Weaver/CompilationFinishedHook.cs

```diff
@@ -35,18 +37,10 @@ static CompilationFinishedHook()
             m_cachedMirrorAssemblyPath = FindMirrorRuntime();
             m_cachedUnityEngineCoreAssemblyPath = UnityEditorInternal.InternalEditorUtility.GetEngineCoreModuleAssemblyPath();
 
-            try
-            {
-                // weave all existing assemblies
-                WeaveAssemblies();
-            }
-            finally
-            {
-                // weave assemblies every time after they are compiled
-                CompilationPipeline.assemblyCompilationFinished += AssemblyCompilationFinishedHandler;
+            // weave assemblies every time after they are compiled
+            CompilationPipeline.assemblyCompilationFinished += AssemblyCompilationFinishedHandler;
```

Comment:
hmmm,  we tried this before and it was causing this issue:

https://github.com/vis2k/Mirror/issues/314

Reply:
It may be the case that Unity changed behaviour, but you may be right. This may be unacceptable behaviour. The order of events if I completely delete my Library folder is:
Unity rebuilds all of Unity's DLLs. Obv it cant call our cctor here because it's not assembled.
Unity calls our cctor
Unity sends us a callback for: Assembly-CSharp.dll and then Assembly-CSharp-Editor.dll
Unity calls our cctor

This brings up some questions, for sure. Why only those two? Surely that's insufficient for many projects. More testing required.

So yeah, what I'll do is take a deep breath (as I'm in no rush to cause a regression for any of our users) and download uMMORPG to test a few things with that. I also would like to make a custom project with various assemblies that reference each other in a dependency chain and confirm Unity's compiler is behaving how I think. These are things worth knowing specifically for this and in general. I'll update the PR once I have fixes for all the issues you've brought up

Reply:
If order is:
Unity calls our cctor
Unity sends us callback for Assembly-CSharp.dll
Unity calls our cctor

then no problem. But the first time it loads,  it appears that the order is:

Unity sends us callback for Assembly-CSharp.dll
Unity calls our cctor

in which case the weaver does not weaver Assembly-CSharp.dll,   According to the linked bug,  it can be reproduced with the examples

Reply:
I don't think that's possible based on the specs of how static constructors aka cctors work in C#. The moment any static member is referenced from a class with a cctor, the runtime automatically calls the cctor right then and there. I bet mono is still playing by the net35 rules where even just instantiating or calling a method invokes the cctor (called eager loading). MS .NET 4.x uses lazy loading where it can wait until a static member is referenced. But in no runtime is it valid to allow static member referencing without cctor being called first, as you are guaranteed a chance to initialize there before being referenced into. Consider if somehow the delegate was invoked without calling the cctor. In my implementation it would result in a null reference from the arrays which aren't yet filled with cached data.

That aside, this is still a problem to my mind. I have a fairly simple solution I'm going to try where we track assembly weave count and ensures each one immediately gets woven if the count is 0. That's the best of both worlds and should solve all known failure cases.

I could not reproduce any issues with the samples because it is indeed weaving Assembly-CSharp.dll, but I can't be sure right at the moment if its doing that on purpose or because of some other conditions which could be variable. I will absolutely find out or solve this a bit differently or both.

This is one of those really tricky situations to debug, because afaik a C# project can't start a process and attach automatically. If you wanted to hit a breakpoint in your InitializeOnLoad static constructor that's what you need. You'd have to add a C++ project within your Unity solutionwith custom debug settings, and even then I'm not sure if it will trigger your breaks from managed code. I think so but I've never tried .. I've only native plugins that way. The other option is you can pause execution in your cctor and use .NET methods to tell when the debugger attaches. But if you do simple Debug.Log statements you will see the cctor is always called  first